docs/src/tutorials/algorithms/tsp_lazy_constraints.jl

```diff
@@ -0,0 +1,326 @@
+# Whenever a subtour has been identified, a constraint corresponding to the form above can be added to the model.
+
+function subtour(edges)
+    ## A list of all unvisited vertices
+    unvisited = Set(collect(1:n))
+
+    ## Placeholder for the shortest subtour
+    cycle = collect(1:n)
+
+    while !(isempty(unvisited))
+        thiscycle = []
+        neighbors = unvisited
+        while !(isempty(neighbors))
+            ## Get the first item         
+            current = pop!(neighbors)
+
+            ## Add it to the current cycle and remove it from unvisited
+            push!(thiscycle, current)
+
+            ## If we are in the first iteration of the inner while loop,
+            ## then the previous pop! already removed `current'
+            if length(thiscycle) > 1
+                pop!(unvisited, current)
+            end
+
+            ## Get the index of all edges to which the current node is connected
+            index = findall(edges -> edges[1] == current, edges)
+
+            ## Based on the index, add the neighbors
+            neighbors = []
+            for i in index
+                append!(neighbors, edges[i][2])
+            end
+
+            ## We only consider neighbors that have not yet been visited
+            neighbors = intersect(neighbors, unvisited)
+        end
+        ## We always store the shortest cycle as subtour
+        if length(thiscycle) < length(cycle)
+            cycle = thiscycle
+        end
+    end
+    return cycle
+end
+
+# Let us declare a helper function `selected_edges()` that will be repeatedly used in what follows.
+
+function selected_edges(x)
+    edges = Tuple{Int,Int}[]
+    for i in 1:n
+        for j in 1:n
+            if (value(x[i, j]) > 0.5)
+                push!(edges, (i, j))
+            end
+        end
+    end
+    return edges
+end
+
+# An iterative way of eliminating subtours is the following.
+# However, it is reasonable to assume that this is not the most efficient way: Whenever a new subtour elimination constraint is added to the model, the optimization has to start from the very beginning.
+# That way, the solver will repeatedly discard useful information encountered during previous solves (e.g., all cuts, the incumbent solution, or lower bounds).
+
+# !!! info
+#     Note that, in principle, it would also be feasible to add all subtours (instead of just the shortest one) to the model.
+#     However, preventing just the shortest cycle is often sufficient for breaking other subtours and will keep the model size smaller.
+
+optimize!(tsp)
+time_iterated = solve_time(tsp);
+cycle = subtour(selected_edges(x))
+
+## Check for cycles, add the subtour elimination constraint,
+## and optimize until no more cycles exist
+while length(cycle) > 1 && length(cycle) < n
+    S = collect(
+        Iterators.filter(t -> t[1] < t[2], Iterators.product(cycle, cycle)),
+    )
```

Comment:
```julia
S = [(i, j) for (i, j) in Iterators.product(cycle, cycle) if i < j]
```

Reply:
Thanks, much cleaner! Fixed in `cf70739`